### PR TITLE
imputation_example,base_impute,error_metrics

### PR DESCRIPTION
**Pull request by cameronmartino**
- Notebook: Imputation Example with 88 soils. 

I chose the 88 soils because it has a higher starting density than the cheese example but the same nice pH trend. 

- error_impute

1. RMSE
2. MSE
3. Frobenius error
4. Density calculation

- base_impute

Simple fill methods: zeros, mean, median, most common in row 

To do:

- [ ] Balance eval metrics

### CODE REVIEW

File: base_impute.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        return imp.transform(A)
+    
+    #most_frequent fill benchmark
+    def most_frequent_fill(A):
+        imp =Imputer(missing_values='NaN', strategy='most_frequent', axis=0, verbose=0, copy=True)
+        imp.fit(A)
+        return imp.transform(A)
+    
+    def xrange(x):
+        return iter(range(x))
```
Comment (mortonjt) [inline base_impute.py]: What's this function for?

File: r_pca.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            ncols = int(sq)
+
+        ymin = np.nanmin(self.D)
+        ymax = np.nanmax(self.D)
+        print('ymin: {0}, ymax: {1}'.format(ymin, ymax))
+
+        numplots = np.min([n, nrows * ncols])
+        plt.figure()
+
+        for n in xrange(numplots):
```
Comment (mortonjt) [inline r_pca.py]: found it!  I think this could be replaced with just range right?

Comment (cameronmartino) [reply r_pca.py]: Confirmed. missed that in the conversion from to py2->3

The x_range function should removed, just had that in there for testing.

**Comment by mortonjt** [comment]
For the heatmaps in code block 63, I'd consider trying

```
from skbio.stats.composition import centralize, clr
sns.heatmap(clr(centralize(observed_table_sni.T)), robust=True,cmap="Blues", ax=ax3)
```

In my experience, this is more effective when contrasting the colors.

There are some style issues that I can submit another PR to this to fix after tomorrow.
We'll want to think a little more carefully about the balance trees benchmark - I can submit another PR to address that. Otherwise, it looks good!

**Comment by cameronmartino** [comment]
I agree, wasn't sure how we should evaluate the balance trees. I just put the dendrogram clustering as a prelude. 

Maybe some kind of unsupervised clustering comparison of silhouette coefficient from sklearn.metrics for the unsupervised tree building when comparing to a known tree ? 

Also add some kind of compositional evaluation metrics ? 

Using  clr/centralize looks great for the starting density OTU but strange for the others. Maybe I am missing something ?

screenshot 2017-02-01 09 54 59

I think it looks better with a color map change and a custom set points? 

sns.heatmap(observed_table.T, robust=False,vmin=0,vmax=2,cmap="Reds", ax=ax1)

red

**Comment by cameronmartino** [comment]
I started a util with a visualization class and a function for plotting variant bacteria over an PCA axis by component weights. Example in the updated notebook.
